### DISCUSSION TITLE
Airflow unable to handle more than 16 tasks at once

### DISCUSSION THREAD

**Discussion by atusman24** [apache/airflow#32609] [Q&A]
Hi,

We have Airflow version 2.5.0 and we're trying to execute a job that contains 60 tasks which can be executed in parallel. Despite trying several concurrency variable values in Airlfow service and environment settings, the airflow is still unable to execute 60 tasks in parallel. 
While executing, it starts out with 32 tasks(screenshot 1,2,3).

<img width="1792" alt="Screenshot 1" src="https://github.com/apache/airflow/assets/131353380/394b6144-6025-4db3-92e6-f89c4b2c6330">
<img width="158" alt="Screenshot 2" src="https://github.com/apache/airflow/assets/131353380/e329eab5-8e70-43ce-9dd6-50cba5d2920d">
<img width="1096" alt="Screenshot 3" src="https://github.com/apache/airflow/assets/131353380/8efa2664-c1a6-476f-b64c-b9d566f98d84">

And after a few moments 16 out of the 32 change status to 'up_for_try' state(screenshot 4,5) and eventually get queued again.

<img width="1169" alt="Screenshot 4" src="https://github.com/apache/airflow/assets/131353380/fa83e057-1372-4645-899c-034cabfca37d">
<img width="178" alt="Screenshot 5" src="https://github.com/apache/airflow/assets/131353380/39d26038-aa8f-42cd-bce3-46bebe32d199">

On the other hand these 32 tasks that are initiated at once are still being executed in ECS as shown in screenshot 6. So it seems like those tasks are still executing while Aiflow is unable to keep track of them and sets them to be rerun. 

<img width="1329" alt="Screenshot 6" src="https://github.com/apache/airflow/assets/131353380/b66d94e4-e8c1-4431-a613-e5d4a2f8dcf2">

Any help would be appreciated. Thank you

**Comment by potiuk** [comment]
A lot depends on resources you have and your deployment. 

You have not explained which executor you have, how many machines you have, what is the parallelism for those machines, what kind of tasks you have them. Airflow as you see - is correclty scheduling the tasks, now it's your job as deployment manager, to make sure that you have enough workers (for celery) with enough resources (CPU/machines, memory, sometimes GPUS etc) to be able to run as many parallel processes you migh want. 

In Kubernetes - you need to have enough resources to be able to run as many PODs you have. 

Depending on the capacity of your system, the number of parallel things that might be executed. It's also your job to make sure that you allocate appropriately the resources (for example via K8S so I suggest you look at your deployment and make sure those resources are not blocking you.

**Reply by atusman24** [reply]
We're using celery executor with the following environment variables:

```
`environment_variables = {
            "AIRFLOW__CORE__SQL_ALCHEMY_CONN": airflow_rds.database_connection_url,
            "AIRFLOW__CORE__EXECUTOR": "CeleryExecutor",
            "AIRFLOW__CORE__DAGS_FOLDER": f"{airflow_efs.mount_point}{airflow_efs.airflow_directory}/dags",
            "AIRFLOW__CORE__DEFAULT_TIMEZONE": "Europe/Berlin",
            "AIRFLOW__CORE__ENABLE_XCOM_PICKLING": "True",
            "AIRFLOW__CORE__DAG_DISCOVERY_SAFE_MODE": "False",
            "AIRFLOW__LOGGING__REMOTE_LOGGING": "True",
            "AIRFLOW__CORE__MAX_ACTIVE_TASKS_PER_DAG": "128",
            "AIRFLOW__CORE__PARALLELISM": "128",
            "AIRFLOW__CORE__DAG_CONCURRENCY": "128",
            "AIRFLOW__CORE__MAX_ACTIVE_RUNS": "128",
            "AIRFLOW__LOGGING__REMOTE_LOG_CONN_ID": "CLOUDWATCH_LOGGING",
            "AIRFLOW__LOGGING__REMOTE_BASE_LOG_FOLDER": f"cloudwatch://{airflow_log_group.log_group_arn}",
            "AIRFLOW__CELERY__BROKER_URL": "sqs://",
            "AIRFLOW__CELERY__DEFAULT_QUEUE": queue.queue_name,
            "AIRFLOW__CELERY__RESULT_BACKEND": f"db+{airflow_rds.database_connection_url}",
            "AIRFLOW__CELERY__CELERY_CONFIG_OPTIONS": "celery_config.CELERY_CONFIG",
            "AIRFLOW__WEBSERVER__DEFAULT_UI_TIMEZONE": "Europe/Berlin",
            "AIRFLOW__WEBSERVER__RBAC": "True",
            "AIRFLOW__WEBSERVER__ENABLE_PROXY_FIX": "True",
            "AIRFLOW__API__AUTH_BACKEND": "airflow.api.auth.backend.basic_auth",
            "AIRFLOW_VAR_USER_DAG_S3_PATH": workspace_bucket.s3_url_for_object(
                "airflow/dags/"
            ),
            "AIRFLOW_VAR_WORKSPACE_BUCKET": workspace_bucket.bucket_name,
            "AIRFLOW_VAR_SQS_QUEUE_NAME": queue.queue_name,
            "AIRFLOW_VAR_SQS_QUEUE_URL": queue.queue_url,
            "AIRFLOW_VAR_ENVIRONMENT": DeploymentSettings.ENVIRONMENT,
            "AIRFLOW_VAR_PROJECT_NAME": DeploymentSettings.PROJECT_NAME,
            "AIRFLOW_VAR_ECS_ENVIRONMENT": json.dumps(
                ecs_environment, separators=(",", ":")
            ),
            "AIRFLOW_VAR_VPC_PRIVATE_SUBNETS": json.dumps(
                vpc_private_subnets, separators=(",", ":")
            ),
            "AIRFLOW_VAR_ALARM_SNS_TOPIC": airflow_alarm_sns_topic.topic_arn,
            "AIRFLOW_VAR_SUCCESS_SNS_TOPIC": airflow_success_sns_topic.topic_arn,
        }`
```

The machine specs are following:

cpu=2048
memory_limit_mib=4096 * 1.5

Also tried with these limits doubled, same outcome.

**Reply by potiuk** [reply]
You need to check how many workers you have, what parallelism you have, how many memory you have, how much memory you have per task and see if those are limiting you. Generally - each task is a separate python interpreter process.

If you have ONE celery worker and you have not changed this parameter:

https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#worker-concurrency

Then yeah. you have maximum 16 parallel celery task processing the queue. So if you need more, you either need to increase that (and make sure your machine can handle it - that's your job as a deployment manager to properly allocate resources and observe your system to see if it is capable and adjust it) or you need to add more workers (i.e. machines) to handle the number of  tasks.

**Reply by atusman24** [reply]
Thank you for the response and pointing out the documentation. I will look into it further to enable concurrency and figure out how I can check and allocate resources that I need to perform all these tasks in parallel. However, I'm still not able to understand why in ECS the 32 tasks initialised through Airflow keep processing while 16 of them are set to 'up_for_retry' and then 'queued' state by Airflow but they keep executing in ECS. To me it seems like ECS keeps executing them and finishes them but then Airflow re-triggers  them again, so in a way some tasks are repeated.

**Reply by potiuk** [reply]
I can't understand either - your explanation does not make it clear. You stil have not explained how many workers and how much concurrency you have, and it is impossible to get into your head as a deployment manager managing your environment and help you.

**Reply by rraghunathan-exelixis** [reply]
@atusman24 are you able to find out why it is queued state? I'm facing the similar issue, im running 40 tasks per dag, but it is only running 23-25 tasks per run, all others are struck in queued state. There is not proper logs on any of the components.

these are my values,
celery:
   worker_concurrency: 50
  core:
    parallelism: 128
    max_active_tasks_per_dag: 50
    max_active_runs_per_dag: 50

Please let me know

**Reply by Peter4net** [reply]
hey, @rraghunathan-exelixis did you figured it out? similar issue here

**Reply by gudata** [reply]
We had the same problem on airflow 2.6 or 2.9 (don't remember)

With a dedicated pool with 256 tasks. I start 50 ec2 machines with enough ram assigned to that pool and only 10 active tasks.

The problem gone when we upgraded to the new airflow 3 without changing anything.

The setup is: celery with airflow in containers from the example compose.yaml

**Reply by rraghunathan-exelixis** [reply]
> hey, @rraghunathan-exelixis did you figured it out? similar issue here

Hey @Peter4net No luck so far. I'm still not able to figure out the resolution. I'm okay with the tasks being in a queued state, but after some time, once the workers have finished their previous execution, they should pick them up — which is not happening. Any luck on your end?
